cmd/juju/commands/bootstrap.go

```diff
@@ -1038,8 +1038,9 @@ func (c *bootstrapCommand) controllerDataRefresher(
 		}
 		addrs = svc.Addresses
 	} else {
-		// TODO(caas): this should never happen.
-		return errors.NewNotValid(nil, "unexpected error happened, IAAS mode should have environs.Environ implemented.")
+		// This should never happen.
+		return errors.New(
+			"supplied BootstrapEnviron implements neither environs.InstanceBroker nor caas.ServiceGetterSetter")
```

Comment:
I usually eschew leaking code-specific terms out in error messages, but in this case it is a programming error and this is the best info that we can emit to diagnose the issue.

Reply:
This should be a panic.